src/Course/StateT.hs

```diff
@@ -267,32 +268,38 @@ instance Functor f => Functor (OptionalT f) where
 instance Monad f => Applicative (OptionalT f) where
   pure = OptionalT . pure . Full
 
-  -- (<*>) :: OptionalT (f (Optional (a -> b))) -> OptionalT (f (Optional a)) -> OptionalT (f (Optional b))
-  -- ffunc :: f (Optional (a -> b))
-  -- fa    :: f (Optional a)
-  -- func <*> :: f _ -> f b
-  OptionalT ffunc <*> OptionalT fa = OptionalT $ (<*>) ffunc <*> _
-
-  -- OptionalT ffunc <*> OptionalT fa = OptionalT $ do
-  --   maybeFunc <- ffunc
-  --   maybeA <- fa
-  --   case maybeFunc of
-  --     Empty -> Empty
-  --     Full func -> case maybeA of
-  --       Empty -> Empty
-  --       Full a -> Full $ func a
-
-
-  -- OptionalT ffunc <*> OptionalT fa = onFull t -> f (Optional a) Optional t
-  --   Applicative f => (t -> f (Optional a)) -> Optional t -> f (Optional a)
-  -- theirs: OptionalT $ (<*>) <$> ffunc <*> fa (but doesn't pass all tests?)
+  OptionalT ffunc <*> OptionalT fa = OptionalT $ do
+    maybeFunc <- ffunc
+    case maybeFunc of
+      Empty     -> pure Empty
+      Full func -> do
+        a <- fa
+        pure $ func <$> a
+    -- REVIEW:
+    -- Above, you suggested solutions like these over more point-free ones like theirs
+    -- (OptionalT f <*> OptionalT a = OptionalT (lift2 (<*>) f a)). Does this fall into the same
+    -- category? I'm still trying to build my intuition around when it makes sense to go point-free
+    -- and when it doesn't.
+    --
+    -- Also, their version fails tests 5-7
+    -- (https://github.com/maxrothman/fp-course/blob/e20fcd61979b76778836dd5277181937ad6facc3/test/Course/StateTTest.hs#L133)
+    -- Mine did too when I had `a <- fa` in the top do. Moving it to the bottom do fixed the issue.
+    --
+    -- Also not really sure what the suggestion to use onFull was about, they don't use it either.
```

Comment:
I don't know what `onFull` is. I can guess that it is probably a catamorphism for `Optional`, i.e. something like `onFull :: b -> (a -> b) -> Optional a -> b`. If so, you can replace the `case` expression with an `onFull` call, but I'm not sure it would look any better.

Reply:
It's defined below: `onFull :: Applicative f => (t -> f (Optional a)) -> Optional t -> f (Optional a)`

Reply:
I don't see how this is different from map

Reply:
It basically just puts a case statement around the second argument

Reply:
It looks exactly like `fmap`, only with the second type parameter specialized to `Optional a`

Reply:
¯\\\_(ツ)_/¯